### PR TITLE
Fix `move_metrics_to_cpu` with evaluation

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes the bug uncovered by the analysis here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10595#issuecomment-972490354

- The outputs are no longer moved to CPU with `move_metrics_to_cpu=True`. This is consistent with how the training loops work.
- The recursive `detach()` call has been removed, it's not necessary as gradients arent computed during evaluation.
- `move_metrics_to_cpu=True` now actually moves the metrics to CPU. This is consistent with how the training loops work.
 
### Does your PR introduce any breaking changes? If yes, please list them.

- If the user relies on the outputs being on CPU in `{validation,test}_epoch_end`.
- If the user memory blows up due to this change. Keep in mind that the user is always able to do `return your_tensor_on_gpu.cpu()`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @edward-io @ananthsub @awaelchli

### CODE REVIEW

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 1]
```diff
@@ -699,3 +699,27 @@ def test_filter_metrics_for_dataloader(kwargs, expected):
...
+
+        def validation_epoch_end(self, outputs):
+            # the step outputs were not moved
+            assert all(o.device == self.device for o in outputs), outputs
+            # but the logging results were
+            assert self.trainer.callback_metrics["foo"].device.type == "cpu"
+
+    model = TestModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir, limit_val_batches=2, move_metrics_to_cpu=True, accelerator="auto", devices=1
```
Comment (rohitgr7) [inline tests/trainer/logging_/test_eval_loop_logging.py:723]: should we run this on just GPU? since it will pass on CPU anyway.

Comment (carmocca) [reply tests/trainer/logging_/test_eval_loop_logging.py]: Done

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 2]
```diff
@@ -134,10 +133,13 @@ def advance(
...
-        if self._should_track_batch_outputs_for_epoch_end():
-            output = recursive_detach(output, to_cpu=self.trainer.move_metrics_to_cpu)
-            if output is not None:
-                self.outputs.append(output)
+        if self._should_track_batch_outputs_for_epoch_end() and output is not None:
+            self.outputs.append(output)
+
+        if self.trainer.move_metrics_to_cpu:
+            # the evaluation step output is not moved as they are not considered "metrics"
+            assert self.trainer._results is not None
```
Comment (justusschock) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:141]: why do we assert this? If somebody doesn't log metrics during validation but during training only and still uses the flag, it would error out here. Why would this be expected?

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 3]
```diff
@@ -134,10 +133,13 @@ def advance(
...
+            self.trainer._results.cpu()
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:142]: This would break if there is reduction performed on epoch end.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:141]: I think it's due to mypy.

Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:141]: Exactly. `Trainer._results` is `Optional` because the predict loop does not define it, but all other loops do

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:142]: You mean as described in https://github.com/PyTorchLightning/pytorch-lightning/issues/10595 right?

It's something to address separately, this PR just makes it works exactly as for training

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:141]: wouldn't

```
if self.trainer._results is not None:
    self.trainer._results.cpu()
```

work?
Seems slightly cleaner than asserting, but as long as we are sure this is never triggered, I'm also fine with the current way

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:141]: I'd prefer that if that was something that could happen, but it's not.

I use assertions because they are short, the other option would be.

```python
if self.trainer.move_metrics_to_cpu:
    if self.trainer._results is None:
        raise InternalError("This shouldn't have happened")
    self.trainer._results.cpu()
```
